### PR TITLE
Feature/actuators unit tests

### PR DESCRIPTION
**Pull request by menglingwei**
add machine unit test cases

### CODE REVIEW

File: pkg/actuators/machine/stubs.go [hunk 1]
```diff
@@ -32,32 +36,90 @@ const (
...
+			userDataSecretKey: []byte(userDataBlob),
+		},
+	}
+}
+
+func stubAlibabaCloudCredentialsSecret() *corev1.Secret {
+	return GenerateAlibabaCloudCredentialsSecretFromEnv(alibabaCloudCredentialsSecretName, defaultNamespace)
+}
+
+func GenerateAlibabaCloudCredentialsSecretFromEnv(secretName, namespace string) *corev1.Secret {
```
Comment (JoelSpeed) [inline pkg/actuators/machine/stubs.go:80]: What's the reason for making this public? Is it being used in other test suites?

File: pkg/actuators/machine/instances_test.go [hunk 2]
```diff
@@ -15,3 +15,1242 @@ limitations under the License.
...
+			// multiple duplicate tags
+			tagList: []machinev1.Tag{
+				{Key: "clusterID", Value: "test-ClusterID"},
+				{Key: "clusterSize", Value: "test-ClusterSize"},
+				{Key: "clusterSize", Value: "test-ClusterSizeDuplicatedValue"},
+			},
+			expected: []ecs.DescribeSecurityGroupsTag{
+				{Key: "clusterID", Value: "test-ClusterID"},
+				{Key: "clusterSize", Value: "test-ClusterSize"},
+				{Key: "clusterSize", Value: "test-ClusterSizeDuplicatedValue"},
```
Comment (JoelSpeed) [inline pkg/actuators/machine/instances_test.go:895]: Are we not expecting this function to filter out duplicate values?

**Review by alexander-demichev — CHANGES_REQUESTED** [review]

File: pkg/actuators/machine/stubs.go (hunk 1 above)
Comment (menglingwei) [reply pkg/actuators/machine/stubs.go:80]: change to generateAlibabaCloudCredentialsSecretFromEnv

File: pkg/actuators/machine/instances_test.go [hunk 3]
```diff
@@ -15,3 +15,1242 @@ limitations under the License.
...
+			// multiple duplicate tags
+			tagList: []machinev1.Tag{
+				{Key: "clusterID", Value: "test-ClusterID"},
+				{Key: "clusterSize", Value: "test-ClusterSize"},
+				{Key: "clusterSize", Value: "test-ClusterSizeDuplicatedValue"},
+			},
+			expected: []vpc.DescribeVSwitchesTag{
+				{Key: "clusterID", Value: "test-ClusterID"},
+				{Key: "clusterSize", Value: "test-ClusterSize"},
+				{Key: "clusterSize", Value: "test-ClusterSizeDuplicatedValue"},
```
[the remark this answers is not in the document]
Comment (menglingwei) [reply pkg/actuators/machine/instances_test.go:936]: i fixed func to filter duplicate key.

File: pkg/actuators/machine/instances_test.go (hunk 2 above)
Comment (menglingwei) [reply pkg/actuators/machine/instances_test.go:895]: i fixed func to filter duplicate key.

File: pkg/actuators/machine/actuator_test.go [hunk 4]
```diff
@@ -38,198 +40,219 @@ import (
...
-func TestMain(m *testing.M) {
-	testEnv := &envtest.Environment{
-		CRDDirectoryPaths: []string{
-			filepath.Join("..", "..", "..", "vendor", "github.com", "openshift", "api", "config", "v1"),
-			filepath.Join("..", "..", "..", "vendor", "github.com", "openshift", "api", "machine", "v1beta1"),
-		},
-	}
-
-	configv1.AddToScheme(scheme.Scheme)
+func Test_Machine(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (menglingwei) [reply pkg/actuators/machine/actuator_test.go]: renamed

File: pkg/actuators/machine/filter_test.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package machine
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+)
+
+func Test_clusterTagFilter(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (menglingwei) [reply pkg/actuators/machine/filter_test.go]: renamed

File: pkg/actuators/machine/instances_test.go [hunk 6]
```diff
@@ -15,3 +15,1242 @@ limitations under the License.
...
+)
+
+func createDescribeInstancesRequest(ids string) *ecs.DescribeInstancesRequest {
+	request := ecs.CreateDescribeInstancesRequest()
+	request.Scheme = "https"
+	request.InstanceIds = ids
+	return request
+}
+
+func Test_runInstances(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (menglingwei) [reply pkg/actuators/machine/instances_test.go]: renamed

File: pkg/actuators/machine/instances_test.go [hunk 7]
```diff
@@ -15,3 +15,1242 @@ limitations under the License.
...
+			} else {
+				if tc.succeeds {
+					t.Errorf("Call to runInstances did not succeed as expected")
+				}
+			}
+		})
+	}
+}
+
+func Test_buildDescribeSecurityGroupsTag(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (menglingwei) [reply pkg/actuators/machine/instances_test.go]: renamed

File: pkg/actuators/machine/instances_test.go [hunk 8]
```diff
@@ -15,3 +15,1242 @@ limitations under the License.
...
+
+	for i, c := range cases {
+		actual := buildDescribeSecurityGroupsTag(c.tagList)
+		if !reflect.DeepEqual(&c.expected, actual) {
+			t.Errorf("test #%d: expected %+v, got %+v", i, c.expected, actual)
+		}
+	}
+}
+
+func Test_buildDescribeVSwitchesTag(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (menglingwei) [reply pkg/actuators/machine/instances_test.go]: renamed

File: pkg/actuators/machine/instances_test.go [hunk 9]
```diff
@@ -15,3 +15,1242 @@ limitations under the License.
...
+			StartTime: "",
+		},
+		{
+			StartTime: "",
+		},
+		{
+			StartTime: time.Now().String(),
+		},
+	}
+	sortInstances(instances)
```
[the remark this answers is not in the document]
Comment (gauravsingh85) [reply pkg/actuators/machine/instances_test.go:1148]: @menglingwei can you reply to Joel comments .

[the remark this answers is not in the document]
Comment (menglingwei) [reply pkg/actuators/machine/instances_test.go:1148]: Sorry, I missed this comments. I updated the test cases.

**Comment by JoelSpeed** [comment]
I think this is good to go, though I'd like to see if I can find a second pair of eyes to review it as well

File: pkg/actuators/machine/machine_scope_test.go [hunk 10]
```diff
@@ -15,3 +15,379 @@ limitations under the License.
...
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	alibabacloudproviderv1 "github.com/openshift/cluster-api-provider-alibaba/pkg/apis/alibabacloudprovider/v1"
+	alibabacloudclient "github.com/openshift/cluster-api-provider-alibaba/pkg/client"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"path/filepath"
+	"reflect"
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (alexander-demichev) [inline pkg/actuators/machine/machine_scope_test.go:36]: this package is imported twice

**Pull request merged**